enhancements/operator-conditions-lib.md

```diff
@@ -0,0 +1,144 @@
+
+1. Read the `Status` CRD:
+
+The library will use controller-runtime's [client][cr_client] to access the resources present in the cluster. To fetch the resource, we will use the [`Get`][cr_get] method of the client, which requires the [`ObjectKey`][cr_objectkey] of type [`types.NamespacedName`][cr_namespacedname]. OLM will set the name of Status CR as an [environment variable][conditions_followup_pr] while creating the resource. Namespace of the operator can be found from the associated [service account secret][accessing_api_from_pod] available in the operators' pod. The library will utilize both of these to obtain the CRD from cluster.
+
+2. Specify conditions in the CRD:
+
+The operators can introduce OLM-supported conditions which influence OLM's behavior or can also specify any custom ([non-OLM supported][non-olm-supported-condition]) condition for communicating the status of the operator to users. An example Status CR is given below:
+
+```yaml
+apiVersion: operators.coreos.com/v1
+kind: Status
+metadata:
+  name: foo-operator-zdgen-asdf # RandomGen Name
+  namespace: operators
+status:
+  conditions:
+    type: Upgradeable # The name of the `foo` OLM Supported Condition.
+    status: "False"   # The operator is not ready to be upgraded.
+    reason: "migration"
+    message: "The operator is performing a migration."
+    lastTransitionTime: "2020-08-24T23:15:55Z"
+```
+
+The operator will be allowed to only modify the [`status`][status-sub-resource] sub-resource of the CR. Operators can either add, delete or update the `status.conditions` array to include OLM-supported or custom conditions. The format and description of the fields present in the conditions can be referred [here][condition_type].
+
+### Scenarios:
+
+1. Operator does not set/update any conditions that affect OLM:
+
+This is a no-op. In this case, OLM will default to its normal behavior.
+
+2. Operator tries to get/update the `Condition` CR when it is not present in the cluster:
+
+In this case, the specific API will return an error. The returned error can be wrapped, such that users can utilize the `errors.Is` method to compare the obtained error with specific values.
+
+3. Operator tries to get/update condition CRD which is present in cluster, but the requested CR instance is not available.
+
+In this case, the error returned will be in the same format as stated in the previous scenario but would specifically mention that the requested object is not found.
+
+To summarize, these are the overall helper methods/functions present in the library. For the ease of understanding, in the below table:
+- `Status` refers to the CR definition provided by OLM and 
+- [`metav1.Condition`][metav1-Condition] refers to specific Condition type present in `status.conditions` array.
+
+ Description             | Input(s)      | Return type(s) |  Expected output  | Remarks    |
```

Comment:
Personally I'd like to see Go function signatures, which are easier to read. Ex:
````md
```go
// SetConditionStatus a metav1.Condition on an apiv1alpha1.Condition object,
// and returns an error if an error scenario is encountered (see Scenarios section).
SetConditionStatus(*apiv1alpha1.Condition, metav1.Condition) error
```
````